policies/wasm/policies/namespace_discovery_relationship_policy/namespace_discovery_relationship_policy.rego

```diff
@@ -1,62 +1,39 @@
-# https://play.openpolicyagent.org/p/wFNhyGsIej
+https://play.openpolicyagent.org/p/iS8qumDa1I
 package hierarchical_policy
 
-available_namespaces[namespace_key] {
-  service := input.services[_];
-  namespace_key = service.namespace
-}
-
-available_namespaces[namespace_key] {
-  service := input.services[_];
-  service.type == "Namespace";
-  namespace_key = service.name
-}
-
-parent_child_mapping = { namespace_map: ns |
-	some namespace_map
-	available_namespaces[namespace_map]
-    service := input.services[key]
-    service.namespace
-    service.namespace == namespace_map
-    ns := {ns_serv: comp_id |
-        some ns_serv
-         x := input.services[ns_serv]
-         x.namespace == namespace_map
-         comp_id := x.traits.meshmap.id
-    }
-}
+import rego.v1
 
-namespaces_to_create[namespaces2] {
-	some namespace
-    available_namespaces[namespace]
-    print(namespace)
-    ns_creation_status = is_present(check_namespace_present_status(namespace))
-    print(ns_creation_status)
-    ns_creation_status != true
-    namespaces2 = namespace
+available_namespaces contains namespace_key if {
+	service := input.services[_]
+	namespace_key := service.namespace
 }
 
-# incase of present: {"present": true}, in absent, returns empty set
-check_namespace_present_status(ns) = is_present {
-  is_present := {"present": is |
-    some svc
-    s := input.services[svc]
-    s.type == "Namespace"
-    s.name == ns
-    is = true
-  }
+available_namespaces contains namespace_key if {
+	service := input.services[_]
+	service.type == "Namespace"
+	namespace_key := service.name
 }
 
-# Is present function wraps the result of check_namespace_present_status to true or false
-is_present(obj) {
-	obj.present == true
+parent_child_mapping[namespace_map] := ns if {
+	some namespace_map in available_namespaces
+	some key, service in input.services
+	service.namespace == namespace_map
+	ns := {ns_serv: comp_id |
+		some ns_serv, x in input.services
+		x.namespace == namespace_map
+		comp_id := x.traits.meshmap.id
+	}
 }
 
-is_present(obj) = pre {
- not obj.present
- pre = false
+namespaces_to_create contains namespace if {
+	some namespace in available_namespaces
+	ns_creation_status := check_namespace_present_status(namespace, input.services)
+	ns_creation_status != true
 }
 
-
-
-
+check_namespace_present_status(ns, all_services) := contains_ns if {
```

Comment:
My assumption based on the existing implementation is that "namespaces_to_create" should end up containing all the namespaces referenced within the input that does not have a matching namespace definition. I used this playground link to test with:

https://play.openpolicyagent.org/p/iS8qumDa1I

So in this example all the pods and services are in the "default" namespace and there is also a "nginx-1" Namespace defined in the list of services. So my assumption was that the "namespaces_to_create" should only contain "default" since there is no explicit "default" namespace defined in the list of services. The existing implementation returned both "default" and "nginx-1". I might be misunderstanding the intention.

Reply:
Nope @anderseknert you are totally correct. It should only return diffs.